content/docs/guides/continuous-delivery/troubleshooting-guide.md

```diff
@@ -0,0 +1,129 @@
+    * For example, GitHub Actions offers the `actions/setup-node` action that can install `node`, `npm`, and `yarn`. Similarly, there is a GitHub Action for installing
+    the runtime for [Python](https://github.com/actions/setup-python), [Go](https://github.com/actions/setup-go), as well as [.NET](https://github.com/actions/setup-dotnet).
+  * If you are using a Docker container-based build, consider using one of Pulumi's [SDK images](https://github.com/pulumi/pulumi/tree/master/docker#sdk-images) for your respective language.
+
+## Restoring Dependencies
+
+Regardless of the language you pick for your Pulumi app, you will have dependencies on at least one other package or library. You must ensure that you
+restore these dependencies before running a `pulumi` command. In the case of `.NET` and `Go` runtimes, those dependencies are automatically restored for you
+when you run `pulumi preview` or `pulumi update --yes`.
+
+### Tips
+
+* For `Node` and `Python` runtimes, you will need to run the step to restore the dependencies ahead of time.
+* For `.NET` and `Go` runtimes, the dependencies are restored for you automatically when you run `pulumi preview` or `pulumi update --yes`.
+* There is an exception to restoring dependencies automatically for `.NET` when you use a private package feed. You must ensure that the
+package(s) from the private feed are accessible or you can use a pre-built binary (see the `binary` option [here]({{< relref "/docs/intro/concepts/project" >}})) with Pulumi to avoid rebuilding your `.NET` solution again.
+  * Note that if you do choose to use a pre-built binary, automatic plugin acquisition for Pulumi
+* You are caching the library dependencies but not the Pulumi plugins. Some services offer dependency caching by capturing a specific folder and restoring
+that folder whenever you run your pipeline subsequently. However, note that Pulumi dependencies have a post-install step that also pulls-down
+a [plugin]({{< relref "/docs/intro/concepts/how-pulumi-works#resource-providers" >}}) binary from our CDN.
+  * So be sure to cache the plugins path as well.
+  * If in doubt about problems encountered during execution, clear out all caches and restore dependencies from scratch.
+
+> Note that depending on the number of providers you use in your Pulumi app, the cache size can get big. Some CI services have
+> restrictions on the max size of the cache.
+
+## Cloud Provider Credentials
+
+Another common point of failure is the cloud provider credentials. The class of errors related to the credentials can be, but not limited to:
+
+* Incorrect credentials (wrong account, mismatched keys etc.)
+* Keys with a strict access scope that don't have access to creating specific resources
+
+### Tips
+
+* Cloud provider credentials not specified using a recognized mechanism. In almost all cases, specifying the credentials through environment variables will work.
```

Comment:
This doesn't read like a tip. Should it start with a "Make sure" or the like? In general, these should probably all read in the second person: "Make sure you've exposed your provider credentials in a supported way", for example, and ideally with a link to what the supported ways are.

Reply:
I will rephrase this sentence to be similar to what you suggest.